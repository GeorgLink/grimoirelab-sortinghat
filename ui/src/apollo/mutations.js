import gql from "graphql-tag";

const LOCK_INDIVIDUAL = gql`
  mutation LockIndividual($uuid: String!) {
    lock(uuid: $uuid) {
      uuid
      individual {
        isLocked
      }
    }
  }
`;

const UNLOCK_INDIVIDUAL = gql`
  mutation UnlockIndividual($uuid: String!) {
    unlock(uuid: $uuid) {
      uuid
      individual {
        isLocked
      }
    }
  }
`;

const DELETE_IDENTITY = gql`
  mutation DeleteIdentity($uuid: String!) {
    deleteIdentity(uuid: $uuid) {
      uuid
    }
  }
`;

const MERGE = gql`
  mutation Merge($fromUuids: [String!], $toUuid: String!) {
    merge(fromUuids: $fromUuids, toUuid: $toUuid) {
      uuid
      individual {
        isLocked
        identities {
          name
          source
          email
          uuid
          username
        }
        profile {
          name
          id
        }
        enrollments {
          organization {
            name
          }
        }
      }
    }
  }
`;

const UNMERGE = gql`
  mutation unmerge($uuids: [String!]) {
    unmergeIdentities(uuids: $uuids) {
      uuids
      individuals {
        profile {
          name
          id
          isBot
        }
        identities {
          name
          source
          email
          uuid
          username
        }
        enrollments {
          start
          end
          organization {
            name
          }
        }
      }
    }
  }
`;

const MOVE_IDENTITY = gql`
  mutation moveIdentity($fromUuid: String!, $toUuid: String!) {
    moveIdentity(fromUuid: $fromUuid, toUuid: $toUuid) {
      uuid
      individual {
        isLocked
        identities {
          name
          source
          email
          uuid
          username
        }
        profile {
          name
          id
        }
        enrollments {
          organization {
            name
          }
        }
      }
    }
  }
`;

const lockIndividual = (apollo, uuid) => {
  let response = apollo.mutate({
    mutation: LOCK_INDIVIDUAL,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

const unlockIndividual = (apollo, uuid) => {
  let response = apollo.mutate({
    mutation: UNLOCK_INDIVIDUAL,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

const deleteIdentity = (apollo, uuid) => {
  let response = apollo.mutate({
    mutation: DELETE_IDENTITY,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

const merge = (apollo, fromUuids, toUuid) => {
  let response = apollo.mutate({
    mutation: MERGE,
    variables: {
      fromUuids: fromUuids,
      toUuid: toUuid
    }
  });
  return response;
};

const unmerge = (apollo, uuids) => {
  let response = apollo.mutate({
    mutation: UNMERGE,
    variables: {
      uuids: uuids
    }
  });
  return response;
};

const moveIdentity = (apollo, fromUuid, toUuid) => {
  let response = apollo.mutate({
    mutation: MOVE_IDENTITY,
    variables: {
      fromUuid: fromUuid,
      toUuid: toUuid
    }
  });
  return response;
};

export {
  lockIndividual,
  unlockIndividual,
  deleteIdentity,
  merge,
  unmerge,
  moveIdentity
};
