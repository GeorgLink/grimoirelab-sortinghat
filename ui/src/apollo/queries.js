import gql from "graphql-tag";

const GET_INDIVIDUAL_BYUUID = gql`
  query GetIndividual($uuid: String!) {
    individuals(filters: { uuid: $uuid }) {
      entities {
        mk
        identities {
          name
        }
        profile {
          id
          name
        }
      }
    }
  }
`;

const GET_INDIVIDUALS = gql`
  query GetIndividuals($page: Int!, $pageSize: Int!) {
    individuals(page: $page, pageSize: $pageSize) {
      entities {
        mk
        isLocked
        identities {
          name
          source
          uuid
        }
        profile {
          id
          name
        }
      }
      pageInfo {
        page
        pageSize
        numPages
        hasNext
        hasPrev
        startIndex
        endIndex
        totalResults
      }
    }
  }
`;

const GET_PROFILE_BYUUID = gql`
  query GetProfileByUuid($uuid: String!) {
    individuals(filters: { uuid: $uuid }) {
      entities {
        isLocked
        profile {
          name
        }
        identities {
          name
          source
          email
          uuid
          username
        }
        enrollments {
          start
          end
          organization {
            name
          }
        }
      }
    }
  }
`;

const getIndividualByUuid = (apollo, uuid) => {
  let response = apollo.query({
    query: GET_INDIVIDUAL_BYUUID,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

const getIndividuals = {
  currentPage: 1,
  numPages: 1,
  pageSize: 10,
  query: async function(apollo, pageSize) {
    let self = this;
    if (pageSize) {
      self.pageSize = pageSize;
    }
    let response = await apollo.query({
      query: GET_INDIVIDUALS,
      variables: {
        page: self.currentPage,
        pageSize: pageSize
      }
    });
    self.numPages = response.data.individuals.pageInfo.numPages;
    if (self.currentPage > self.numPages) {
      return undefined;
    }
    self.currentPage++;
    return response;
  }
};

const getProfileByUuid = (apollo, uuid) => {
  let response = apollo.query({
    query: GET_PROFILE_BYUUID,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

export { getIndividuals, getIndividualByUuid, getProfileByUuid };
