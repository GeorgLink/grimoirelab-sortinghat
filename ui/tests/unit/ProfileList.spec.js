import { shallowMount, mount, createLocalVue } from "@vue/test-utils";
import Vue from "vue";
import Vuetify from "vuetify";
import ProfileList from "@/components/ProfileList";
import * as Queries from "@/apollo/queries";

Vue.use(Vuetify);

const profileMocked = {
  data: {
    individuals: {
      entities: [{
        isLocked: false,
        profile: {
          name: "Test Name"
        },
        identities: [{
          name: "Test Name",
          source: "github",
          email: "test@example.net",
          uuid: "7a727227226daae693c61ddf7040e51c97ac638d",
          username: "test"
        }],
        enrollments: []
      }],
      __typename: "IdentityPaginatedType"
    }
  }
};

describe("ProfileList", () => {
  test("Mock query for getProfileByUuid", async () => {
    const query = jest.fn(() => Promise.resolve(profileMocked));
    const wrapper = shallowMount(ProfileList, {
      Vue,
      mocks: {
        $apollo: {
          query
        }
      },
      computed: {
        selectedIndividual: () => {
          return {
            uuid: "12abc",
            selected: true
          }
        }
      }
    });
    const response = await Queries.getProfileByUuid(wrapper.vm.$apollo, wrapper.vm.selectedIndividual.uuid);
    expect(query).toBeCalled();
    expect(wrapper.element).toMatchSnapshot();
  });

  test("Queries profile data when selected", async () => {
    const getProfileByUuidSpied = spyOn(Queries, "getProfileByUuid");
    const query = jest.fn(() => Promise.resolve(profileMocked));
    const wrapper = shallowMount(ProfileList, {
      Vue,
      attachToDocument: true,
      mocks: {
        $apollo: {
          query
        }
      },
      computed: {
        selectedIndividual: () => {}
      }
    });

    wrapper.vm.$options.watch.selectedIndividual.call(wrapper.vm, {
      uuid: "12abc",
      selected: true
    });

    expect(getProfileByUuidSpied).toHaveBeenCalledWith(wrapper.vm.$apollo, "12abc");
  });

  test("Shows selected profile", async () => {
    const localVue = createLocalVue();
    const vuetify = new Vuetify();
    const query = jest.fn(() => Promise.resolve(profileMocked));
    const wrapper = mount(ProfileList, {
      Vue,
      vuetify,
      mocks: {
        $apollo: {
          query
        }
      },
      data() {
        return {
          profiles: []
        }
      },
      computed: {
        selectedIndividual: () => {}
      }
    });

    wrapper.vm.$options.watch.selectedIndividual.call(wrapper.vm, {
      uuid: "12abc",
      selected: true
    });

    await Vue.nextTick();
    expect(wrapper.vm.profiles.length).toBe(1);
    await Vue.nextTick();
    expect(wrapper.find('.v-navigation-drawer--open').exists()).toBe(true);
    expect(wrapper.findAll('.v-list-item__title').at(1).text()).toBe("Test Name");
    expect(wrapper.element).toMatchSnapshot();
  });
});
