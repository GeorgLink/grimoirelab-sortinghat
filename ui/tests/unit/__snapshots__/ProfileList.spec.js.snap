// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileList Mock query for getProfileByUuid 1`] = `
<v-navigation-drawer-stub
  color="#eceff1"
  fixed="true"
  height="100%"
  min-height="calc(100vh - 66px)"
  minivariantwidth="56"
  mobilebreakpoint="1264"
  right="true"
  src=""
  style="top: 64px; z-index: 1;"
  tag="aside"
  width="626"
>
   
</v-navigation-drawer-stub>
`;

exports[`ProfileList Shows selected profile 1`] = `
<aside
  class="v-navigation-drawer v-navigation-drawer--fixed v-navigation-drawer--is-mobile v-navigation-drawer--open v-navigation-drawer--right theme--light"
  min-height="calc(100vh - 66px)"
  style="height: 100%; top: 64px; transform: translateX(0%); width: 626px; background-color: rgb(236, 239, 241); border-color: #eceff1; z-index: 1;"
>
  <div
    class="v-navigation-drawer__prepend"
  >
    <div
      class="v-list-item theme--light"
      tabindex="-1"
    >
      <div
        class="v-list-item__content"
      >
        <div
          class="v-list-item__title"
        >
          Selected profiles
        </div>
      </div>
    </div>
     
    <hr
      aria-orientation="horizontal"
      class="v-divider theme--light"
      role="separator"
    />
  </div>
  <div
    class="v-navigation-drawer__content"
  >
     
    <div
      class="row"
      style="padding: 10px 0px;"
    >
      <div
        class="mx-auto v-card v-card--link v-card--raised v-sheet theme--light"
        style="width: 600px;"
        tabindex="0"
      >
        <div
          class="grow v-list-item theme--light"
          tabindex="-1"
        >
          <div
            class="v-avatar v-list-item__avatar grey"
            style="height: 40px; min-width: 40px; width: 40px;"
          >
            
      TN
    
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__title"
            >
              Test Name
            </div>
             
            <div
              class="v-list-item__subtitle"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-github theme--light"
                style="font-size: 16px;"
              />
            </div>
          </div>
           
          <div
            class="v-list-item__icon"
          >
            <button
              class="v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-lock-open-outline theme--light"
                  style="font-size: 16px;"
                />
              </span>
            </button>
          </div>
        </div>
         
        <div
          class="v-list-group"
        >
          <div
            aria-expanded="false"
            class="v-list-group__header v-list-item v-list-item--link theme--light"
            role="button"
            tabindex="0"
          >
            <div
              class="v-list-item__title"
            >
              
        Identities (1)
      
            </div>
            <div
              class="v-list-item__icon v-list-group__header__append-icon"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-chevron-down theme--light"
              />
            </div>
          </div>
          <!---->
        </div>
         
        <div
          class="v-list-group"
        >
          <div
            aria-expanded="false"
            class="v-list-group__header v-list-item v-list-item--link theme--light"
            role="button"
            tabindex="0"
          >
            <div
              class="v-list-item__content"
            >
              <div
                class="v-list-item__title"
              >
                
          Organizations (0)
        
              </div>
            </div>
            <div
              class="v-list-item__icon v-list-group__header__append-icon"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-chevron-down theme--light"
              />
            </div>
          </div>
          <!---->
        </div>
         
        <!---->
      </div>
    </div>
  </div>
  <div
    class="v-navigation-drawer__border"
  />
</aside>
`;
